test/reline/test_key_actor_emacs.rb

```diff
@@ -1243,13 +1243,14 @@ def test_ed_search_prev_history_with_empty
     ])
     # The ed_search_prev_history doesn't have default binding
     @line_editor.__send__(:ed_search_prev_history, "\C-p".ord)
-    assert_line_around_cursor('', '12345')
+    assert_line_around_cursor('12345', '')
     @line_editor.__send__(:ed_search_prev_history, "\C-p".ord)
-    assert_line_around_cursor('', '12aaa')
+    assert_line_around_cursor('12345', '')
```

Comment:
This part should be `('12aaa', '')`
#618 might be harder than I first thought.

I checked GNU Readline.
```
ruby -rreadline -rreline -e "raise if Readline==Reline; loop{p Readline.readline 'prompt> ', true}"
```

```
prompt> █
ed_search_prev_history (search with empty string)
prompt> 12345█
ed_search_prev_history (A) (search with empty string)
prompt> 12aaa█
```

```
prompt> █
ed_search_prev_history (search with empty string)
prompt> 12345█
Move cursor left and right. Continuous search_xxx_history is reset.
prompt> 12345█
ed_search_prev_history (B) (search with "12345")
prompt> 12345█
```
There is a difference between (A) and (B). Same input, same cursor position, but there is a hidden state. We need to introduce a new state (instance variable) to achieve this.

Reply:
Adding new instance variable for many `def ed_[key_action_name]` will make maintainability poor, so I think we should introduce a general-purpose feature for passing state to next action.

One idea is:
```ruby
def input_key
  ...
  # process_key() or normal_char() is called here
  ...
  @prev_action_state, @next_action_state = @next_action_state, nil
end
```

Key action method can read `@prev_action_state`.
If the action want to pass a state to next action, set `@next_action_state` to something like `[state_type, state_value]`.

Reply:
Thank you for your clear comment. To replicate the behavior of Readline, it is best to include information about the before and after states of the current operation. I will give it a try

Reply:
fixed in 7ddd20d7184ec35cc35805c5d7d937620221c023